consensus.go

```diff
@@ -195,18 +200,35 @@ func New(validator SignKey, transport Transport, opts ...ConfigOption) *Pbft {
 		state:        newState(),
 		transport:    transport,
 		msgQueue:     newMsgQueue(),
-		updateCh:     make(chan struct{}),
+		updateCh:     make(chan struct{}, 100), //hack. There is a bug when you have several messages pushed on the same time.
```

Comment:
We use in PushInternal such code:

```code	

select {

case p.updateCh <- struct{}{}:

default:

}```

that notifies getNextMessage that we have a new message. 

If we haven't read the previous message from p.updateCh we'll fall to default and won't get an update.  I've just added a buffer to make it not reproducible. It may be not a problem from v3(Need to be investigated).

Reply:
@b00ris not sure have you seen explanation for such pattern on task you have created (https://linear.app/matic/issue/POS3-183/missing-update).
I'm going to paste explanation here as well, just to have conversation streamlined:

We have previously discussed with @ferranbt about part of code used here and were asking the same question. 🙂 

However, as Ferran explained, the thing is that `updateCh` is used just as a trigger, which signals to start next iteration of messages reading loop in the `getNextMessage` function. Therefore we don't need to act upon each message being pushed into the queue, as we are going to read it eventually (or discard if we accumulate enough messages for the given round and sequence to proceed to the next Pbft state).

Reply:
This would be my only minor concern here. It is okay if we refactor the rest of the property-based toolkit as another PR but I am not sure this behaves correctly as it is right now with the buffer.

Reply:
@Stefan-Ethernal @ferranbt I've checked it again. Tests become probabilistic if i remove capacity.
I'm not sure that I got your idea. As i understand updateCh used as a trigger to process one more message. There is a function that reads messages.
```
func (p *Pbft) getNextMessage(span trace.Span) (*MessageReq, bool) {
	for {
		msg, discards := p.notifier.ReadNextMessage(p)
...
		// wait until there is a new message or
		// someone closes the stopCh (i.e. timeout for round change)
		select {
		case <-p.state.timeoutChan:
...
		case <-p.ctx.Done():
			return nil, false
		case <-p.updateCh:
		}
}
```
For example: 
Cluster ABCD. A should receive 3 Prepare messages.
So we need to call PushMessage for B, C, D. It adds a message to a queue and trigger updateCh.
```
func (p *Pbft) PushMessageInternal(msg *MessageReq) {
	p.msgQueue.pushMessage(msg)

	select {
	case p.updateCh <- struct{}{}:
	default:// we miss trigger here
	}
}
```
A receives B, C and D at the same time(trigger once).
A read B and process, then read  trigger from updateCh, read C and got stuck because there is no trigger for the last message. 
We have to wait for timeout message to exit getNextMessage cycle.
Or we have to guarantee that pbft.PushMessage can't be called concurrent. Does my explanation correlate with the result of your discussion?

Reply:
But the thing here is that `getNextMessage` shouldn't proceed to `select-case` block if there is a message for the current `sequence` (block height) and `round`. Only if it don't find appropriate message it waits either for new message (`updateCh` notification), timeout or cancellation to happen in order to again try to read message for current `sequence` and `round`. I mean pbft algorithm shouldn't get stuck, since we are having timeouts as well and in that case, if timeout elapses, it will jump out from `getNextMessage`.

Reply:
Yes, messages are correct. One node could stuck if it receive the message after ReadNextMessage but before select. I think the validator will and wait before timeout or next message that push to updateCh or receive created block from another peer.

Reply:
> One node could stuck if it receive the message after ReadNextMessage but before select.

If message is valid (namely for the current `sequence` and `round`), it is jumped out of the `getNextMessage` immediately after it is read and execution won't event fall down to `select-case`. Am I missing out something here? 

Are you talking here about use case where message queue is either empty or contains only messages for some past sequences/rounds (i.e. only messages that are going to be discarded). In such case, I believe that we could miss out some message notification, because of that `default` statement in `PushMessageInternal` and in that case timeout would occur. However, message is still injected to the queue and there is chance that it could be read or discarded on some other occasion. Is my understanding correct?

Reply:
The main problem with this thing is: to push in a not buffered channel we need someone who read from it at the same time(or we'll switch to default section). The simplest solution is to make channel buffered.